File: src/main/scala/com/spark3d/geometryObjects/Point3D.scala
/*
 * Copyright 2018 Julien Peloton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.astrolabsoftware.spark3d.geometryObjects

import scala.collection.JavaConverters._

import com.astrolabsoftware.spark3d.geometryObjects.Shape3D._
import com.astrolabsoftware.spark3d.utils.Utils.sphericalToCartesian

/**
  * Class for describing a point in 3D space.
  * By default, the input coordinates are supposed euclidean, that is (x, y, z).
  * The user can also work with spherical input coordinates (x=r, y=theta, z=phi)
  * by setting the argument isSpherical=true.
  *
  * @param x : (Double)
  *   Input X coordinate in Euclidean space, and R in spherical space.
  * @param y : (Double)
  *   Input Y coordinate in Euclidean space, and THETA in spherical space.
  * @param z : (Double)
  *   Input Z coordinate in Euclidean space, and PHI in spherical space.
  * @param isSpherical : (Boolean)
  *   If true, it assumes that the coordinates of the Point3D are (r, theta, phi).
  *   Otherwise, it assumes cartesian coordinates (x, y, z).
  *
  */
class Point3D(val x: Double, val y: Double, val z: Double,
    val isSpherical: Boolean) extends Shape3D with Serializable {

  // The center of the point is the point
  val center: Point3D = this

  // Zero volume
  def getVolume: Double = 0.0

  /**
    * Methods to determine whether the Point3D is contained in another shape.
    * Implement different ways for different shapes (Point, Shell, Box available).
    *
    * @param otherShape : (Shape3D)
    *   An instance of Shape3D (or extension)
    * @return (Boolean) true if the two objects intersect.
    *
    */
  override def intersects(otherShape: Shape3D): Boolean = {

    // Different methods to handle different shapes
    if (otherShape.isInstanceOf[Point3D]) {
      otherShape.asInstanceOf[Point3D].isEqual(this)
    } else if (otherShape.isInstanceOf[ShellEnvelope]) {
      otherShape.asInstanceOf[ShellEnvelope].isPointInShell(this)
    } else if (otherShape.isInstanceOf[BoxEnvelope]) {
      otherShape.asInstanceOf[BoxEnvelope].contains(this)
    } else {
      throw new AssertionError(
        """
        Cannot perform intersection because the type of shape is unknown!
        Currently implemented:
          - point x point
          - point x sphere
          - point x box
        """)
    }
  }

  /**
    * Get bounding box of this Point3D which will also be a Point3D.
    * Perform a conversion spherical to euclidean if needed.
    *
    * @return (BoxEnvelope) Cuboid representing the box of the Point3D
    */
  override def getEnvelope(): BoxEnvelope = {
    if (this.isSpherical) {
      new BoxEnvelope(sphericalToCartesian(this))
    } else {
      new BoxEnvelope(this)
    }

  }

  /**
    * Returns the distance between the point and another.
    * The two points must have the same coordinate system.
    * Space is supposed flat (euclidean).
    *
    * @param p : (Point3D)
    *   Another instance of Point3D
    * @return (Double) Distance between the two points.
    *
    */
  def distanceTo(p: Point3D): Double = {
    if (p.isSpherical != this.isSpherical) {
      throw new AssertionError("""
        The 2 points must have the same coordinate system to compute
        the distance! Convert one using sphericalToCartesian or
        cartesianToSpherical methods.
        """)
    }

    val module = if (!this.isSpherical) {
      math.sqrt(
        (this.x - p.x) * (this.x - p.x) +
          (this.y - p.y) * (this.y - p.y) +
          (this.z - p.z) * (this.z - p.z)
      )
    } else {
      math.sqrt(
        this.x * this.x + p.x * p.x - 2 * this.x * p.x * (
          math.sin(this.y) * math.sin(p.y) *
            math.cos(this.z - p.z) +
            math.cos(this.y) * math.cos(p.y)
          )
      )
    }
    module
  }

  /**
    * Return the coordinates (x, y, z) of the point.
    *
    * @return (List[Double]) The list of coordinates.
    */
  def getCoordinate: List[Double] = {
    List(this.x, this.y, this.z)
  }

  /**
    * Python interface for getCoordinate.
    * py4j automatically converts Java List into readable python list,
    * but struggle with Scala List.
    * Return the coordinates (x, y, z) of the point.
    *
    * @return (java.util.List[Double]) The list of coordinates.
    */
  def getCoordinatePython: java.util.List[Double] = {
    List(this.x, this.y, this.z).asJava
  }

  /**
    * Return if the input Point3D is equal this Point3D
    *
    * @param p (Point3D)
    *   Point3D for which the equality is to be checked
    * @return (Boolean) true if the two Point3Ds are equal
    */
  def isEqual(p: Point3D): Boolean = {
    x == p.x &&
    y == p.y &&
    z == p.z
  }
}

File: pyspark3d/geometryobjects/point3d.py
# Copyright 2018 Julien Peloton
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from pyspark import SparkContext
from pyspark3d import load_from_jvm
from pyspark3d_conf import pyspark3d_conf

import os
import doctest
import numpy as np

def Point3D(x: float, y: float, z: float, isSpherical: bool):
    """
    Class to handle Point3D

    Parameters
    ----------
    x : float
        Input X coordinate in Euclidean space, and R in spherical space.
    y : float
        Input Y coordinate in Euclidean space, and THETA in spherical space.
    z : float
        Input Z coordinate in Euclidean space, and PHI in spherical space.
    isSpherical : bool
        If true, it assumes that the coordinates of the Point3D
        are (r, theta, phi). Otherwise, it assumes cartesian
        coordinates (x, y, z).

    Example
    ----------
    >>> p3d = Point3D(1.0, np.pi, 0.0, True)
    >>> print(type(p3d))
    <class 'py4j.java_gateway.JavaObject'>

    >>> p3d.getVolume()
    0.0

    >>> p3d.toHealpix(2048, True)
    50331644
    """
    scala_p3d = "com.astrolabsoftware.spark3d.geometryObjects.Point3D"
    p3d = load_from_jvm(scala_p3d)

    return p3d(x, y, z, isSpherical)


if __name__ == "__main__":
    """
    Run the doctest using

    python __init__.py

    If the tests are OK, the script should exit gracefuly, otherwise the
    failure(s) will be printed out.
    """
    # Activate the SparkContext for the test suite
    pwd = os.environ["PWD"]
    spark3d_jar = os.path.join(
        pwd, "../../target/scala-2.11/spark3d_2.11-0.1.5.jar")
    healpix_jar = os.path.join(
        pwd, "../../lib/jhealpix.jar")
    sparkfits_maven = "com.github.astrolabsoftware:spark-fits_2.11:0.6.0"

    dic = {
        "spark.jars": spark3d_jar+","+healpix_jar,
        "spark.jars.packages": sparkfits_maven
        }

    conf = pyspark3d_conf("local", "test", dic)
    sc = SparkContext.getOrCreate(conf=conf)

    # Numpy introduced non-backward compatible change from v1.14.
    if np.__version__ >= "1.14.0":
        np.set_printoptions(legacy="1.13")

    # Run the test suite
    doctest.testmod()
